### ISSUE TITLE
instance not included

### ISSUE THREAD

**Issue by fpvandoorn** [leanprover/lean3#1007]
In the following example, the instance `[is_homomorphism φ]` is not included in the theorem `respect_one`

``` lean
-- standard
import algebra.group

open algebra

definition is_homomorphism [class] {G₁ G₂ : Type} [group G₁] [group G₂] (φ : G₁ → G₂) : Prop :=
Π(g h : G₁), φ (g * h) = φ g * φ h

variables {G₁ G₂ : Type} (φ : G₁ → G₂) [group G₁] [group G₂] [is_homomorphism φ]

theorem respect_one : φ 1 = 1 := _
```

**Comment by leodemoura** [comment]
@fpvandoorn `is_homomorphism` is tagged as a `[class]`. Is this a typo?

**Comment by fpvandoorn** [comment]
No, that was intentional. The notation `[foo X]` is only useful if `foo` is a class (or reduces to one), right?

**Comment by fpvandoorn** [comment]
There is the question whether you would want this definition to be a class. I (tentatively) think it is useful.

**Comment by leodemoura** [comment]
I see what is the problem.
The current rule for including an "annonymous instance implicit argument" is:
- All parameters of the instance have been used in the declaration.

`[is_homomorphism φ]` depends on `[group G₁] [group G₂]`, and they don't appear explicitly in the declaration. To address this issue, we have to change the condition above.
- All parameters of the instance have been used in the declaration  _or have been included as anonymous instances_

**Comment by fpvandoorn** [comment]
Yes, I think that is a better rule.

**Comment by robertylewis** [comment]
I just noticed something while thinking about this issue. I'm not sure if this has always been the case or if it changed since the last time I looked....

In the algebraic hierarchy, we often have to `include` instances for them to be recognized. See, e.g., https://github.com/leanprover/lean/blob/master/library/algebra/group.lean#L315 -- without this line, the following theorems fail. However, sometimes we don't have to do this: see https://github.com/leanprover/lean/blob/master/library/algebra/group.lean#L148 . Is there an obvious difference here that I'm missing? The failure to automatically include the `add_group` instance doesn't seem to be a case of what Floris points out.
